pyvista/core/utilities/geometric_sources.py

```diff
@@ -15,40 +15,65 @@
 from .helpers import wrap
 
 
-def translate(surf, center=(0.0, 0.0, 0.0), direction=(1.0, 0.0, 0.0)):
+def translate(
+    surf, center=(0.0, 0.0, 0.0), direction=(1.0, 0.0, 0.0), start_direction=(1.0, 0.0, 0.0)
+):
     """Translate and orient a mesh to a new center and direction.
 
-    By default, the input mesh is considered centered at the origin
-    and facing in the x direction.
+    The input mesh is considered centered at the origin
+    and facing in the start_direction.
 
     Parameters
     ----------
     surf : pyvista.core.pointset.PolyData
         Mesh to be translated and oriented.
-    center : tuple, optional, default: (0.0, 0.0, 0.0)
+    center : tuple, default: (0.0, 0.0, 0.0)
         Center point to which the mesh should be translated.
-    direction : tuple, optional, default: (1.0, 0.0, 0.0)
+    direction : tuple, default: (1.0, 0.0, 0.0)
         Direction vector along which the mesh should be oriented.
+    start_direction : tuple, default: (1.0, 0.0, 0.0)
+        Direction vector along which mesh is originally oriented.
 
```

Comment:
I might argue that the current test already does this because we reuse the exact machinery for `start_direction`.  We are now doing two rotations to get the right overall rotation. First from `start_direction` to `(1, 0, 0)` then from `(1, 0, 0)` to `direction`. The result is matrix multiplied.  The first rotation is calculated as the transpose of `(1, 0, 0)` to `start_direction`, which is already tested except for the transpose which is numerically robust.  And we have separately tested that the transform using `start_direction` is correct.  But, having said all that, it was quite simple to add in a test, and now we have it in case anyone in the future cleans up this code to derive the rotation matrix without doing two separate rotations.

Reply:
Actually, I'm not sure that the test I added is exactly relevant to what you wrote.  Specifically I tested `start_direction=(0, 1+eps, 0)` and the negative version. Because of the way I implemented this, the same numerical issues should occur with `start_direction` as `direction`.  Since there weren't any tests for `direction=(1+eps, 0, 0)`, I'm assuming we don't need any for `start_direction`.  But I don't have a good feeling for how to know what other combinations can lead to numerical instability.

Reply:
So the main instabilities came from these lines: https://github.com/pyvista/pyvista/blob/6093145c1fdef4304c7bce7d4eccd31b7eabbb72/pyvista/core/utilities/geometric_sources.py#L44-L53

Without the `allclose()` checks we could end up with the cross product of near-collinear vectors, which would be close to a zero vector, which has close to 0 length, so dividing by its norm (when normalising `normz`) was the source of instability I think.

The current implementation mostly uses the (hopefully) fixed version of `transpose()`, so we only have to look at the new code path. On this I see https://github.com/pyvista/pyvista/blob/6093145c1fdef4304c7bce7d4eccd31b7eabbb72/pyvista/core/utilities/geometric_sources.py#L60-L67

The potential issue I see here is the exact comparison of two vectors. So what I'd test is when `start_direction` is _almost_ the same as `direction`, and when `start_direction` is almost _equal_ to `[1, 0, 0]` (these are the two exact comparisons). The algo is not at all obvious to me so I won't try to figure out after a long day what the expected result would be in this case :) But this is the kind of instability I can imagine here.

Reply:
Thanks for continuing to poke at this piece.  I refactored it assuming that the changes made sense in the first place.  Taking a closer look, this function does something unexpected to me.  This function rotates around the y-axis, i.e. `[0, 1, 0]`, in all cases except when `direction` is close to `[0, 1, 0]`, then it rotates around the z-axis. This will lead to situations where small variations in direction will cause significantly large changes in rotation.  I think we should have instead raised an error or logged a warning.  I need to dig through this a little more...

Reply:
Mathematically speaking, rotating from one axis to the other is not a well-defined operation (because if you imagine an x-y-z local coordinate system, rotating the `z` axis into an arbitrary `v` vector still gives you freedom of the orientation of the other two axes around `v`). Algorithmically speaking there's an (I think) unambiguous choice of always rotating along the great circle connecting the start and the end direction. You can do this by computing the cross product ot start and end (which will be perpendicular to the plane defined by the start and end vectors) and then rotating around that vector. I'd naively implement this using [Rodrigues' rotation formula](https://en.wikipedia.org/wiki/Rodrigues%27_rotation_formula).  You'd only have to check cases when the two vectors are close to collinear, in which case the cross products go to zero.

I _suspect_ the current algorithm gives the same result, but I don't understand what it does by just looking at it, and I haven't had the time to write down on paper what those multiple cross products do. I suspect it essentially generates all three vectors of the rotated coordinate system (containing `v` in my example above), and for this you need multiple orthonormal vectors, and this is typically when one tries to choose an arbitrary vector (`[0, 1, 0]`) to generate orientations with.

Long story short:

> After more testing, it seems that the result is robust and doesn't result in large changes in rotation with small changes in direction around `[0, 1, 0]`, although I haven't wrapped my head around it yet. The current code logic was mostly to prevent machine precision level inaccuracies to build up during the computation rather than to account for some instability. I think I was overthinking this a bit and just need to add tests for `start_direction` equals `direction` and almost equals. And the same for `start_direction` equals `(1, 0, 0)` and almost equals `(1, 0, 0)`.

This sounds reasonable to me in light of my earlier comment :) Thanks for your patience.

Reply:
Yes, I was going to point out the same general feature about the non-uniqueness of the rotations that rotate one vector onto another, and then yes, also, that the great circle approach using the cross product and Rodrigues formula seems like the way to go. I *also* can't tell if the current approach is doing this.

However, all of that said

(1) Even with the above algorithm, an arbitrary choice needs to be made about what to do if the two vectors are exactly anti-parallel. Then there is no canonical choice of axis to rotate around.
(2) Also with the above algorithm: If the start and target vectors are *close* to anti-parallel then there will be a sort of instability where tiny changes in the target vector will result it largely different rotation matrices.

My suggestion: If the user really wants to rotate an object in 3D space they needs to give a better specification of the rotation. Either by passing Euler angles or a full blown rotation matrix or something. I guess the `direction` feature is inheriting from the same feature on the `Sphere` function (which maybe inherits from some VTK stuff?). My opinion is that while, on the surface, this feature looks a nice convenience, it's kind of under-specified and will result in weird behaviors even when it's working as expected. As a user, I would rather do the work on constructing the rotation matrix I want on my own and then either rotate the mesh myself (in which case the pyvista functions should have no functionality for rotating the meshes they create) or pass my rotation matrix into the function and have that used to perform the rotation (in which case the `direction` parameter should be replaced by a `rotation_matrix` or `rot_mat` parameter).

I think my short term suggestion would be to remove the `direction` parameter from the `SolidSphere` functions for now and get the PR through. Then, in a separate PR, visit the questions of whether the `direction` parameter makes sense to add to `SolidSphere` and/or the question of if the `direction` parameter on `Sphere` should actually be modified as well to something that gives the user better control over the rotation. Maybe a backward compatible path forward could be to include a `rot_mat` parameter on `SolidSphere` and then on `Sphere` to keep the `direction` parameter (for backwards compatibility, (maybe with a deprecation path?) but also add the `rot_mat` parameter as a more sensible rotation specification.

This all just my 2 cents on the rotation stuff. Maybe taken with a grain of salt since I haven't poured into the existing code too much yet.

Reply:
I'm more and more leaning to reverting these changes and figure out how to manually rotate the `SolidSphere` to be consistent with `Sphere`.

Reply:
@MatthewFlamm That seems fine to me. I'd be fine with either

1. Matching `direction` behavior for `SolidSphere` to the old behavior of `Sphere` (i.e. not modifying the behavior of `Sphere` `direction` at all and matching `SolidSphere` to that old behavior.
Or
2. Excluding the `direction` parameter from `SolidSphere` entirely.

Basically doing any work rethinking how the `direction` parameter works feels out of scope for this PR.